File: cliente.cpp
#include "cliente.h"

Cliente::Cliente(int _rol)
{
    connect(&cliente, SIGNAL(connected()),
      this, SLOT(enviar()));

    rol = _rol;

    if(rol == 1)
    {
        servidor_cliente = new Servidor(8080); // 8080 mientras sirve el input del usuario
        servidor_cliente->start();
        //Eliminar luego
        archivo = "hola mundo";
        enlistar_paquetes();
        tamano_ventana = 3;
        timeout=1;
        //
    }
}

/*
 * Destructor para el cliente
*/
Cliente::~Cliente()
{
  cliente.close();
}

/*
 * Conecta al cliente con un puerto
*/
void Cliente::conectar(QString ip, quint16 puerto)
{
  QHostAddress direccion_ip(ip);
  cliente.connectToHost(direccion_ip, puerto);
}

/*
 * Compara el tiempo del inicio del primer paquete enviado con el tiempo actual
*/
bool Cliente::timeout_alcanzado(double tiempo_paquete, clock_t ahora)
{
    double duracion = ( ahora - tiempo_paquete ) / (double) CLOCKS_PER_SEC;
    if(duracion >= timeout)
    {
        return true;
    } else {
        return false;
    }
}

/*
 * Toma el QString "archivo" que contiene la frase a enviar y la parsea en paquetes,
 * posteriormente encola esos paquetes en la cola de paquetes.
*/
void Cliente::enlistar_paquetes()
{
    for(int i=0; i<archivo.length(); i++)
    {
        paquete *nuevo_paquete = new paquete(-1, '0', std::clock(), true);
        nuevo_paquete->colocar_secuencia(i);
        nuevo_paquete->colocar_valor(archivo.at(i));
        //
        qDebug() << "Encolando: " << nuevo_paquete->obtener_secuencia() << ":" << nuevo_paquete->obtener_valor();
        //
        cola_de_paquetes.push_back(nuevo_paquete);
    }
}

/*
 * Forma el paquete que será enviado
*/
void Cliente::ensamblar_paquete(int indice)
{
    paquete_a_enviar="";
    //Ensambla el paquete a enviar
    paquete_a_enviar = QString::number(cola_de_paquetes[indice]->obtener_secuencia()) + ":";
    paquete_a_enviar = paquete_a_enviar + cola_de_paquetes[indice]->obtener_valor();
    cola_de_paquetes[indice]->comenzar_timer();
}

/*
 * Envían data al puerto al que se conecta el cliente
*/
void Cliente::enviar()
{
    if(rol==1 && id==QThread::currentThreadId())
    {
        //manda el paquete
        qDebug() << "[Cliente:" << this->rol << "]" << "Ahora enviando: " << paquete_a_enviar;
        const char *enviar_str = paquete_a_enviar.toStdString().c_str();
        cliente.write(enviar_str, paquete_a_enviar.length());
    }
}

/*
 * Comienza la ejecución del hilo del cliente
*/
void Cliente::run()
{

    this->id = QThread::currentThreadId();
    qDebug() << "[Cliente] : El cliente " << this->rol << " está corriendo ahora.";
    if(rol==1)
    {
        int indice_ultimo_archivo_enviado;
        //Espera hasta que haya un archivo a enviar
        while(archivo.isEmpty())
        {
            //Esperando a que el usuario seleccione un archivo con la interfaz
        }
        //Envia los primeros "tamaño ventana" paquetes
       if(!cola_de_paquetes.empty())
        {
            int tope_de_primer_envio = tamano_ventana;
            //Caso en el que la ventana sea más grande que el archivo a enviar.
            if(tamano_ventana > archivo.length())
            {
                tope_de_primer_envio = archivo.length();
            }
            indice_ultimo_archivo_enviado = tope_de_primer_envio;
            for(int i=0; i<tope_de_primer_envio; i++)
            {
                ensamblar_paquete(i);
                emit enviar();
            }
            qDebug() << "Envié los primeros tamano_ventana paquetes";
        }
        //Ciclo principal que se encarga de enviar todos los paquetes
        while(!cola_de_paquetes.empty())
        {
            qDebug() << "----------------------Ciclo principal--------------------";
            while(servidor_cliente->lecturas_vacia()
                  && !timeout_alcanzado(cola_de_paquetes.front()->obtener_reloj(), std::clock()))
            {
                //El tiempo pasa
            }
            //si se venció el timeout
            if(servidor_cliente->lecturas_vacia())
            {
                qDebug() << "Se venció el timeout";
                for(int i=0; i<indice_ultimo_archivo_enviado; i++) //antes tamano ventana
                {
                    ensamblar_paquete(i);
                    emit enviar(); //Aqui
                }
            //si se recibió el ACK
            } else {
                while(!servidor_cliente->lecturas_vacia())
                {
                    QByteArray ack_leido = servidor_cliente->obtener_ultima_lectura();
                    if(ack_leido.toInt() == cola_de_paquetes.front()->obtener_secuencia())
                    {
                        //TODO imprimir "Llegó ACK de cola paquetes first.seq"
                        cola_de_paquetes.erase(cola_de_paquetes.begin());
                        //Si aun quedan paquetes, deslizo la ventana
                        if(!cola_de_paquetes.empty())
                        {
                            indice_ultimo_archivo_enviado++;
                            ensamblar_paquete(indice_ultimo_archivo_enviado);
                            emit enviar();
                        }
                    }
                }
            }
        }
    }
}

File: Listo para pruebas/src/cliente/Cliente.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package cliente;

import java.util.*;
import java.io.*;
import java.net.*;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.lang.*;

/**
 *
 * @author jean
 */
public class Cliente {

    Servidor servidor = new Servidor();
    LinkedList<Paquete> cola_de_paquetes = new LinkedList<Paquete>();
    int tamano_ventana = 3;
    String archivo = "Melissa";
    int puerto_intemediario;
    Boolean modo;
    long timeout = 1000;
    String paquete_a_enviar;

    public Boolean timeout_alcanzado(long tiempo_paquete, long ahora) {
        long duracion = (ahora - tiempo_paquete);
        return duracion >= timeout;
    }

    public void enlistar_paquetes() {
        for (int i = 0; i < archivo.length(); i++) {
            Paquete nuevo_paquete = new Paquete(-1, '0', System.currentTimeMillis());
            nuevo_paquete.secuencia = i;
            nuevo_paquete.valor = archivo.charAt(i);
            System.out.println("Encolando: " + nuevo_paquete.secuencia + ":" + nuevo_paquete.valor);
            cola_de_paquetes.addLast(nuevo_paquete);
        }
    }

    public void ensamblar_paquete(int indice) {
        paquete_a_enviar = "";
        //Ensambla el paquete a enviar
        paquete_a_enviar = cola_de_paquetes.get(indice).secuencia
                + ":";
        paquete_a_enviar = paquete_a_enviar + cola_de_paquetes.get(indice).valor;
        cola_de_paquetes.get(indice).reloj = System.currentTimeMillis();
    }

    public void enviar() {
        try {
            Socket clientSocket = new Socket("localhost", 6789);
            DataOutputStream outToServer
                    = new DataOutputStream(clientSocket.getOutputStream());
            outToServer.writeBytes(paquete_a_enviar + '\n');
            clientSocket.close();
        } catch (IOException ex) {
            Logger.getLogger(Cliente.class.getName()).log(Level.SEVERE, null, ex);
        }
    }

    public void iniciar() {
        servidor.start();
//
        System.out.println("[Cliente] : El cliente está corriendo ahora.");
        int tope_a_enviar=tamano_ventana;
        //Espera hasta que haya un archivo a enviar
        while (archivo.isEmpty()) {
            //Esperando a que el usuario seleccione un archivo con la interfaz
        }
        enlistar_paquetes();
        //Envia los primeros "tamaño ventana" paquetes
        if (!cola_de_paquetes.isEmpty()) {
            int tope_de_primer_envio = tamano_ventana;
            //Caso en el que la ventana sea más grande que el archivo a enviar.
            if (tamano_ventana > archivo.length()) {
                tope_de_primer_envio = archivo.length();
            }
            for (int i = 0; i < tope_de_primer_envio; i++) {
                ensamblar_paquete(i);
                enviar();
            }
        }
        //Ciclo principal que se encarga de enviar todos los paquetes
        while (!cola_de_paquetes.isEmpty()) {
            while(servidor.cola_de_lecturas.isEmpty()
                    && !timeout_alcanzado(cola_de_paquetes.element().reloj, System.currentTimeMillis()))
            {
                //El tiempo pasa
            }
            //si se venció el timeout
            if(servidor.cola_de_lecturas.isEmpty())
            {
                System.out.println("Se venció el timeout\n");
                for(int i=0; i<tope_a_enviar; i++) 
                {
                    ensamblar_paquete(i);
                    enviar();
                }
            //si se recibió el ACK
            } else {
                while(!servidor.cola_de_lecturas.isEmpty())
                {
                    String ack_leido = servidor.cola_de_lecturas.pop();
                    if(ack_leido.equals(Integer.toString(cola_de_paquetes.element().secuencia)))
                    {
                        System.out.println("Llego el ack: " + ack_leido);
                        cola_de_paquetes.pop();
                        //Si aun quedan paquetes, deslizo la ventana
                        if(!cola_de_paquetes.isEmpty())
                        {
                            if(cola_de_paquetes.size() < tamano_ventana)
                            {
                                tope_a_enviar=cola_de_paquetes.size();
                            } else {
                                ensamblar_paquete(tamano_ventana - 1);
                                enviar();
                            }
                        } 
                    }
                }
            }
        }
    }
}
